artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -210,6 +211,9 @@
 
    private final Runnable deliverRunner = new DeliverRunner();
 
+   //This lock is used to prevent deadlocks between direct and async deliveries
+   private final ReentrantLock deliverLock = new ReentrantLock();
```

Comment:
Ouch really??? Seriouslly!!

Reply:
That was happening because MQTT callback wasn't implemented to support direct delivery, but was wrongly  reporting to use it. 
Anyway using this lock will prevent direct deliveries to happen concurrently with an async deliveries, that is a way to enforce a correct behaviour anyway.
Not sure if thanks to this MQTT could enable direct deliveries, probably not.

Reply:
Point is we now have two different lock approaches in this class. Recipe for disaster

Reply:
Not sure: I have made explicit an implicit requirement. And now the order of locking is ensured just into QueueImpl , while before it was enforced in different classes (callbacks, transactions and queue).

Reply:
Re: mqtt i don't see why it can't have direct deliveries if CORE can, either we have a locking issue in code or we don't.